data/transactions/verify/verifiedTxnCache.go

```diff
@@ -0,0 +1,292 @@
+
+const entriesPerBucket = 8179 // the default bucket size; a prime number could promote a lower hash collisions in case the hash function isn't perfect.
+const maxPinnedEntries = 500000
+
+// VerifiedTxnCacheError helps to identifiy the errors of a cache error and diffrenciate these from a general verification errors.
+type VerifiedTxnCacheError struct {
+	inner error
+}
+
+// Unwrap provides accesss to the underlying error
+func (e *VerifiedTxnCacheError) Unwrap() error {
+	return e.inner
+}
+
+// Error formats the underlying error message
+func (e *VerifiedTxnCacheError) Error() string {
+	return e.inner.Error()
+}
+
+// errTooManyPinnedEntries is being generated when we attempt to pin an transaction while we've already exceeded the maximum number of allows
+// transactions in the verification cache.
+var errTooManyPinnedEntries = &VerifiedTxnCacheError{errors.New("Too many pinned entries")}
+
+// errMissingPinnedEntry is being generated when we're trying to pin a transaction that does not appear in the cache
+var errMissingPinnedEntry = &VerifiedTxnCacheError{errors.New("Missing pinned entry")}
+
+// VerifiedTransactionCache provides a cached store of recently verified transactions
+type VerifiedTransactionCache interface {
+	// Add adds a given transaction group and it's associated group context to the cache. If any of the transactions already appear
+	// in the cache, the new entry overrides the old one.
+	Add(txgroup []transactions.SignedTxn, groupCtx *GroupContext)
+	// AddPayset works in a similar way to Add, but is intended for adding an array of transaction groups, along with their corresponding contexts.
+	AddPayset(txgroup [][]transactions.SignedTxn, groupCtxs []*GroupContext) error
+	// GetUnverifiedTranscationGroups compares the provided payset against the currently cached transactions and figure which transaction groups aren't fully cached.
+	GetUnverifiedTranscationGroups(payset [][]transactions.SignedTxn, CurrSpecAddrs transactions.SpecialAddresses, CurrProto protocol.ConsensusVersion) [][]transactions.SignedTxn
+	// UpdatePinned replaces the pinned entries with the one provided in the pinnedTxns map. This is typically expected to be a subset of the
+	// already-pinned transactions. If a transaction is not currently pinned, and it's can't be found in the cache, a errMissingPinnedEntry error would be generated.
+	UpdatePinned(pinnedTxns map[transactions.Txid]transactions.SignedTxn) error
+	// Pin function would mark the given transaction group as pinned.
+	Pin(txgroup []transactions.SignedTxn) error
+}
+
+// verifiedTransactionCache provides an implementation of the VerifiedTransactionCache interface
+type verifiedTransactionCache struct {
+	// bucketsLock is the lock for syncornizing the access to the cache
+	bucketsLock deadlock.Mutex
+	// buckets is the circular cache buckets buffer
+	buckets []map[transactions.Txid]*GroupContext
+	// pinned is the pinned transactions entries map.
+	pinned map[transactions.Txid]*GroupContext
```

Comment:
seems like `pinned` is used as a kind of a cache within the cache for faster lookup in `GetUnverifiedTranscationGroups`. Could please put a comment here and maybe in the interface about pinning?

Reply:
not quite. It's not about faster lookup; it's about non-perishable by incoming transactions. I'll add some comments to explain that.